### PR TITLE
DBZ-7920 Add support for 'infinity' values for PostgreSQL

### PR DESCRIPTION
**Pull request by mfvitale**
closes: https://issues.redhat.com/browse/DBZ-7920

Missing task:
- [x] Perform a manual e2e test
- [x] Add support and e2e tests for other destination platforms
    - [x] MySQL
    - [x] Oracle
    - [x] SQLServer
    - [x] Db2

### CODE REVIEW

**Comment by Naros** [comment]
Hi @mfvitale , one question.  I see you marked do manual e2e test, what about these values from PG going to other database platforms?  Should we introduce an E2E test that only applies to the PG source but can be run against the other sink targets for completeness?

**Comment by mfvitale** [comment]
> what about these values from PG going to other database platforms? 

This is a good question. I just tried PostgreSQL to PostgreSQL. I'll add e2e for the others.

**Comment by Naros** [comment]
So I suspect we likely will want to add some methods on the `DatabaseDialect` so that various implementations can control what values should be serialized to the target system in the case of `+/-INFINITY` from PostgreSQL:

It looks like for Oracle, MySQL, SQL Server, and Db2, they all map as follows:

|Infinity Constant|Date/Time Value|
|---|---|
|`-INFINITY`|`1000/01/01 00:00:00`|
|`+INFINITY`|`9999/12/31 23:59:59`|

**Comment by mfvitale** [comment]
> So I suspect we likely will want to add some methods on the `DatabaseDialect` so that various implementations can control what values should be serialized to the target system in the case of `+/-INFINITY` from PostgreSQL:
> 
> It looks like from what I quickly found:
> 
> Vendor	-INFINITY	+INFINITY
> Oracle/MySQL	`1000/01/01 00:00:00`	`9999/12/31 23:59:59`
> SQL Server/Db2	`0001/01/01 00:00:00`	`9999/12/31 23:59:59`

Thanks for it. 

This open another question: should we manage also the opposite? These values should be converted to '+/-infinity' when PostgreSQL is the destination?

**Comment by mfvitale** [comment]
This is my final founding about the other databases support

|Vendor|-INFINITY|+INFINITY|
|---|---|---|
|Oracle|`-4712-01-01T00:00:00+00:00`|`9999-12-31T23:59:59+00:00`|
|MySQL|`1970-01-01T00:00:01+00:00`|`2038-01-19T03:14:07+00:00`|
|SQL Server/Db2|`0001/01/01 00:00:00`|`9999/12/31 23:59:59`|

**Comment by mfvitale** [comment]
@Naros can you give another look?

**Comment by Naros** [comment]
> This open another question: should we manage also the opposite? These values should be converted to `+/-infinity` when PostgreSQL is the destination?

That's a good question and I guess it would depend on the complexity.  I haven't looked at your recent changes, but I could imagine that if PostgreSQL calendar limits less than the other database vendors, then yes that would make logical sense to imply +/- INFINITY in that case.

File: src/main/java/io/debezium/connector/jdbc/dialect/db2/ZonedTimestampType.java [hunk 1]
```diff
@@ -23,34 +18,28 @@
  *
  * @author Chris Cranford
  */
-public class ZonedTimestampType extends AbstractTimestampType {
+public class ZonedTimestampType extends io.debezium.connector.jdbc.type.debezium.ZonedTimestampType {
```
Comment (Naros) [inline src/main/java/io/debezium/connector/jdbc/dialect/db2/ZonedTimestampType.java:21]: I must admit, I'm not a fan of extending a class with the same name; should we think about prefixing the derived class with a database prefix of sorts?

File: src/test/java/io/debezium/connector/jdbc/e2e/AbstractJdbcSinkPipelineIT.java [hunk 2]
```diff
@@ -36,6 +36,7 @@
 
 import org.apache.commons.lang3.RandomStringUtils;
 import org.apache.kafka.connect.sink.SinkRecord;
+import org.jetbrains.annotations.NotNull;
```
Comment (Naros) [inline src/test/java/io/debezium/connector/jdbc/e2e/AbstractJdbcSinkPipelineIT.java:39]: Lets remove this

File: src/test/java/io/debezium/connector/jdbc/e2e/AbstractJdbcSinkPipelineIT.java [hunk 3]
```diff
@@ -2568,6 +2569,51 @@ else if (sink.getType().is(SinkType.MYSQL)) {
...
+                values,
+                expectedValues,
+                (record) -> {
+                    assertColumn(sink, record, "data0", getTimestampWithTimezoneType(source, false, 6));
+                    assertColumn(sink, record, "data1", getTimestampWithTimezoneType(source, false, 6));
+                },
+                (rs, index) -> rs.getTimestamp(index).toInstant().atZone(ZoneOffset.UTC));
+    }
+
+    private static @NotNull List<ZonedDateTime> getExpectedZonedDateTimes(Sink sink) {
```
Comment (Naros) [inline src/test/java/io/debezium/connector/jdbc/e2e/AbstractJdbcSinkPipelineIT.java:2593]:
```suggestion
    private static List<ZonedDateTime> getExpectedZonedDateTimes(Sink sink) {
```

**Review by Naros — COMMENTED** [review]
LGTM, just a few inline comments & my comment above about translating to PG +/- INFINITY values.

File: src/main/java/io/debezium/connector/jdbc/dialect/db2/ZonedTimestampType.java (hunk 1 above)
Comment (mfvitale) [reply src/main/java/io/debezium/connector/jdbc/dialect/db2/ZonedTimestampType.java:21]: Or we can prefix the base one with "Debezium" as done for "Connect" one. WDYT? I am open for both.

File: src/test/java/io/debezium/connector/jdbc/e2e/AbstractJdbcSinkPipelineIT.java (hunk 2 above)
Comment (mfvitale) [reply src/test/java/io/debezium/connector/jdbc/e2e/AbstractJdbcSinkPipelineIT.java:39]: Yes, I missed it.

**Comment by mfvitale** [comment]
> I haven't looked at the recent changes, but I could imagine that if the calendar limits for another vendor go beyond the limits of PostgreSQL, then it would make logical sense to convert those to +/- INFINITY since we would be incapable of writing the actual date value into PostgreSQL without an error.

The only issue seems to be only for the Oracle -infinity value (-4712-01-01T00:00:00+00:00). In that case maybe it is better to convert all infinity values from the source to the correct destination value. So I think it is better to manage this in a separate issue. WDYT?

File: src/main/java/io/debezium/connector/jdbc/dialect/db2/ZonedTimestampType.java (hunk 1 above)
Comment (Naros) [reply src/main/java/io/debezium/connector/jdbc/dialect/db2/ZonedTimestampType.java:21]: > Or we can prefix the base one with "Debezium" as done for "Connect" one.

Ya lets go with that approach, its more consistent with what we have already with prefixes.

**Comment by Naros** [comment]
> The only issue seems to be only for the Oracle -infinity value (-4712-01-01T00:00:00+00:00). In that case maybe it is better to convert all infinity values from the source to the correct destination value. So I think it is better to manage this in a separate issue. WDYT?

Makes sense to me. :+1:

**Review by Naros — APPROVED** [review]
Reran the tests for Oracle and they all checked out :+1:, so LGTM.

**Comment by mfvitale** [comment]
> > The only issue seems to be only for the Oracle -infinity value (-4712-01-01T00:00:00+00:00). In that case maybe it is better to convert all infinity values from the source to the correct destination value. So I think it is better to manage this in a separate issue. WDYT?
> 
> Makes sense to me. 👍

Opened https://issues.redhat.com/browse/DBZ-8033

**Pull request merged**